File: app/utils/dawproject/referenceable.ts
import { Nameable } from "./nameable";
import { IReferenceable } from "./types";
import { Utility } from "./utility";

export abstract class Referenceable extends Nameable implements IReferenceable {
  /** Unique string identifier of this element. This is used for referencing this instance from other elements. */
  id: string;
  private static ID_COUNTER = 0;
  private static _instances: { [id: string]: Referenceable } = {};

  /** call before export */
  static resetIdCounter(): void {
    Referenceable.ID_COUNTER = 0;
    Referenceable._instances = {};
  }

  constructor(name?: string, color?: string, comment?: string) {
    super(name, color, comment);
    this.id = `id${Referenceable.ID_COUNTER++}`;
    Referenceable._instances[this.id] = this;
  }

  toXmlObject(): any {
    const attributes = super.toXmlObject(); // get attributes from Nameable

    // add required attribute
    Utility.addAttribute(attributes, "id", this, {
      required: true,
    });

    return attributes;
  }

  fromXmlObject(xmlObject: any): this {
    super.fromXmlObject(xmlObject); // populate inherited attributes from Nameable

    // validate and populate required attribute
    Utility.populateAttribute(xmlObject, "id", this, {
      required: true,
    });

    // register instance for reference lookup
    Referenceable._instances[this.id] = this;

    return this;
  }

  static getById(id: string): Referenceable | undefined {
    return Referenceable._instances[id];
  }
}

File: app/utils/dawproject_python/referenceable.py
from classes.nameable import Nameable


class Referenceable(Nameable):
    ID = 0
    _instances = {}

    @classmethod
    def reset_id(cls):
        cls.ID = 0
        cls._instances = {}

    def __init__(self, name=None, color=None, comment=None):
        super().__init__(name, color, comment)
        self.id = f"id{Referenceable.ID}"
        Referenceable._instances[self.id] = self
        Referenceable.ID += 1

    def to_xml(self):
        element = super().to_xml()
        element.set("id", self.id)
        return element

    @classmethod
    def from_xml(cls, element):
        instance = super().from_xml(element)
        instance.id = element.get("id")
        cls._instances[instance.id] = instance
        return instance

    @classmethod
    def get_by_id(cls, id):
        return cls._instances.get(id)
